src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,968 @@
+                            { cell, v_id }, vertices[v_id] );
+                    }
+                }
+                else if( nb_faces == 5 )
+                {
+                    const index_t cell = mesh_builder->create_cells(
+                        (index_t) 1, CellType::PYRAMID );
+
+                    std::vector< index_t > vertices = {
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[1],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[0],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[3],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[2],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 1 )[0]
+                    };
+
+                    for( auto v_id : range( 5 ) )
+                    {
+                        mesh_builder->set_cell_vertex(
+                            { cell, v_id }, vertices[v_id] );
+                    }
+                }
+                else if( nb_faces == 6 )
+                {
+                    const index_t cell = mesh_builder->create_cells(
+                        (index_t) 1, CellType::HEXAHEDRON );
+
+                    std::vector< index_t > vertices = {
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[0],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[1],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[3],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 0 )[2],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 1 )[0],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 1 )[1],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 1 )[3],
+                        (index_t) unstructed_grid.getNodeIndicesOfFaceOfCell(
+                            cell, 1 )[2]
+                    };
+
+                    for( auto v_id : range( 8 ) )
+                    {
+                        mesh_builder->set_cell_vertex(
+                            { cell, v_id }, vertices[v_id] );
+                    }
+                }
+            }
+            unstructed_grid.unloadGeometry();
+            mesh_builder->connect_cells();
+
+            return true;
+        }
+    } // namespace
+
+    bool GeoModelBuilderRESQMLImpl::read_volumes( const EpcDocument& pck )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 46 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)